examples/sst2/input_pipeline.py

```diff
@@ -0,0 +1,271 @@
+# Copyright 2020 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""SST-2 input pipeline."""
+
+import collections
+from typing import Dict, List, Set, Text, Tuple
+import os
+
+from absl import logging
+
+import numpy as np
+import tensorflow.compat.v2 as tf
+import tensorflow_datasets as tfds
+
+
+# pylint: disable=no-member
+
+
+# pylint: disable=abstract-method
+class OrderedCounter(collections.Counter, collections.OrderedDict):
+  """A counter that remembers the order of added keys."""
+  pass
+
+
+def get_tokens(datasets: List[tf.data.Dataset]) -> Set[Text]:
+  """Returns a set with all unique tokens in the given datasets."""
+  counter = OrderedCounter()
+  for ds in datasets:  # Data set already tokenized here.
+    for example in tfds.as_numpy(ds):
+      tokens = whitespace_tokenize(example['sentence'].strip())
+      counter.update(tokens)
+  sst_tokens = set(counter.keys())
+  logging.info('Number of unique tokens: %d', len(sst_tokens))
+  return sst_tokens
+
+
+def save_vocab(vocab, path):
+  """Saves a vocabulary to disk."""
+  with open(path, mode='wb') as f:
+    for token in vocab:
+      f.write(token + b'\n')
+  logging.info('Saved vocab to: %s', path)
+
+
+def load_vocab(path):
+  """Loads a vocabulary from disk."""
+  vocab = collections.OrderedDict()
+  with open(path, mode='rb') as f:
+    for i, token in enumerate(f):
+      vocab[token.rstrip(b'\n')] = i
+  return vocab
+
+
+def save_embed(embed, path):
+  """Save word embeddings to disk."""
+  assert path.endswith('.npz'), 'Path must end with .npz'
+  np.savez(path, embed=embed)
+  logging.info('Saved embed to: %s', path)
+
+
+def load_embed(path):
+  """Load word embeddings from disk."""
+  return np.load(path)['embed']
+
+
+# pylint: disable=dangerous-default-value
```

Comment:
I guess this is safe since "specials" is not modified in the function. However, it is a bit dangerous practice: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

Maybe add a note that this is safe since we don't modify "specials"?

Reply:
Fixed